### PR TITLE
Validate with `-testUsedAssemblies` in a CI leg

### PR DESCRIPTION
**Pull request by jcouv**
This identified a few issues tests that assert or throw when `GetEmitDiagnostics()` is called: [VoidEvent](https://github.com/dotnet/roslyn/issues/39979), [NullCoalescingOperatorWithTupleNames](https://github.com/dotnet/roslyn/issues/60046), [ExternAliases4](https://github.com/dotnet/roslyn/issues/60045), [CrossAssemblySupportingAndNotSupportingCovariantReturns](https://github.com/dotnet/roslyn/issues/60060), [TestNullCoalesce_Dynamic](https://github.com/dotnet/roslyn/issues/60059)

Verified the validation fails as expected with a [canary](https://dev.azure.com/dnceng/public/_build/results?buildId=1657429&view=ms.vss-test-web.build-test-results-tab).

Fixes https://github.com/dotnet/roslyn/issues/60086

### CODE REVIEW

File: eng/build.ps1 [hunk 1]
```diff
@@ -374,6 +376,10 @@ function TestUsingRunTests() {
     $env:ROSLYN_TEST_IOPERATION = "true"
   }
 
+  if ($testEmitDiagnostics) {
+    $env:ROSLYN_TEST_EMITDIAGNOSTICS = "true"
+  }
+
   if ($testUsedAssemblies) {
```
Comment (jcouv) [inline eng/build.ps1:383]: @AlekseyTs Should we enable `testUsedAssemblies` in a CI leg? Seems like it's unused at the moment.

Comment (AlekseyTs) [reply eng/build.ps1:383]: > Should we enable testUsedAssemblies in a CI leg? Seems like it's unused at the moment.

I think it would be good to add a separate CI leg similar to the one used for IOperation testing.

Comment (jcouv) [reply eng/build.ps1:383]: Okay. Let's keep that separate (there's a few questions, such as choosing a leg, how to deal with potential failures that might have snuck in). 
Filed tracking [issue](https://github.com/dotnet/roslyn/issues/60086).

Towards that goal, this PR fixes an issue with TestRunner that makes sure we copy environment variables to Helix machines where tests end up running.

**Comment by jcouv** [comment]
@cston @RikkiGibson @dotnet/roslyn-compiler for review. Thanks

File: eng/build.ps1 (hunk 1 above)
Comment (RikkiGibson) [reply eng/build.ps1:383]: I think we should consider whether to group this set of "additional validation" flags into a single leg or to have separate legs. Recently we are trying to be sensitive about number of test legs introduced because:

1. Each time we additionally run the tests increases the rate of spurious failures.
2. Cost of provisioning the additional machines.

See #57313.

**Review by RikkiGibson — APPROVED** [review]
Implementation looks good, but perhaps consideration from the team is required to decide whether to do this validation in the existing job or in a new job.

**Comment by jcouv** [comment]
FYI @jmarolf From discussion with Rikki and team, we're adding a new CI leg (debug, core, compiler tests only, with some additional validation added to every compiler test for used assemblies and emit diagnostics).

File: azure-pipelines.yml [hunk 2]
```diff
@@ -120,6 +120,15 @@ jobs:
     configuration: Debug
     testArguments: -testCoreClr -testIOperation -testCompilerOnly
 
+template: eng/pipelines/test-windows-job.yml
```
Comment (jaredpar) [inline azure-pipelines.yml:123]: Consider adding a comment here explaining why this job is diff than others.

File: eng/build.ps1 [hunk 3]
```diff
@@ -465,14 +465,12 @@ function TestUsingRunTests() {
...
 
+    # Note: remember to update TestRunner when using new environment variables
+    # (they need to be transferred over to the Helix machines that run the tests)
     if ($testIOperation) {
       Remove-Item env:\ROSLYN_TEST_IOPERATION
     }
 
-    if ($testUsedAssemblies) {
-      Remove-Item env:\ROSLYN_TEST_USEDASSEMBLIES
-    }
```
Comment (jaredpar) [inline eng/build.ps1:474]: Why was this removed?

File: azure-pipelines.yml [hunk 4]
```diff
@@ -120,6 +120,17 @@ jobs:
     configuration: Debug
     testArguments: -testCoreClr -testIOperation -testCompilerOnly
 
+# This leg runs almost all the compiler tests supported on CoreCLR, but
+#   with additional validation for used assemblies and GetEmitDiagnostics
```
Comment (jaredpar) [inline azure-pipelines.yml:124]:
```suggestion
#  with additional validation for used assemblies and GetEmitDiagnostics
```

File: src/Compilers/CSharp/Test/Emit/PDB/PDBUsingTests.cs [hunk 5]
```diff
@@ -569,7 +569,7 @@ class A { void M() {  } }
 ");
         }
 
-        [Fact]
+        [ConditionalFact(typeof(NoUsedAssembliesValidation))] // Tracked by https://github.com/dotnet/roslyn/issues/60045
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Emit/PDB/PDBUsingTests.cs:572]: Why do you use comments here vs. putting the URL into the `Reason` property?

**Review by jaredpar — APPROVED** [review]

File: src/Compilers/CSharp/Test/Emit/PDB/PDBUsingTests.cs (hunk 5 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Emit/PDB/PDBUsingTests.cs:572]: Not much reason. Either one works

Comment (333fred) [reply src/Compilers/CSharp/Test/Emit/PDB/PDBUsingTests.cs:572]: One appears in the test logs, one does not 🙂.

File: src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs [hunk 6]
```diff
@@ -4823,7 +4823,9 @@ public static D Dynamic_b_not_null_a_nullable(int? c)
     }
 }
 ";
-            var compilation = CreateCompilation(source).VerifyDiagnostics();
+            // We use VerifyDiagnosticsOnly until this issue is fixed:
```
[the remark this answers is not in the document]
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs]: I agree that skipped and conditional tests could drift, but conditional tests on `NoUsedAssembliesValidation` and `NoIOperationValidation` are how we deal with problematic tests in those CI legs at the moment. I'd rather not revisit that for purpose of this PR

**Review by RikkiGibson — APPROVED** [review]
